File: src/conduit/loader/snowflake_loader.py
"""Snowflake destination writer."""

from __future__ import annotations

import logging
from pathlib import Path

import pyarrow as pa

from conduit.models import DestinationConfig

logger = logging.getLogger(__name__)


def load_snowflake(table: pa.Table, dest: DestinationConfig, base_dir: Path) -> None:
    """Write an Arrow table to a Snowflake table."""
    try:
        import snowflake.connector
        from snowflake.connector.pandas_tools import write_pandas
    except ImportError:
        raise ImportError(
            "snowflake-connector-python[pandas] is required for Snowflake destinations. "
            "Install it with: pip install snowflake-connector-python[pandas]"
        )

    conn_params = {
        "account": dest.config.get("account"),
        "user": dest.config.get("user"),
        "password": dest.config.get("password"),
        "warehouse": dest.config.get("warehouse"),
        "database": dest.config.get("database"),
        "schema": dest.config.get("schema", "PUBLIC"),
        "role": dest.config.get("role"),
    }
    conn_params = {k: v for k, v in conn_params.items() if v is not None}

    target_table = dest.config.get("table", dest.name).upper()

    conn = snowflake.connector.connect(**conn_params)
    try:
        cur = conn.cursor()

        if dest.mode == "full_refresh":
            cur.execute(f"DROP TABLE IF EXISTS {target_table}")

        # Convert Arrow → pandas → Snowflake via write_pandas
        df = table.to_pandas()
        write_pandas(
            conn,
            df,
            table_name=target_table,
            auto_create_table=True,
            chunk_size=dest.batch_size,
        )
        logger.info(
            "Loaded %d rows to Snowflake destination '%s' (%s)",
            len(table), dest.name, target_table,
        )
    finally:
        conn.close()

File: src/ts/loader/snowflake_loader.ts
/**
 * Snowflake destination writer.
 */

// deno-lint-ignore-file no-explicit-any
import snowflake from "snowflake-sdk";
import type { DataTable, DestinationConfig } from "../models.ts";

function execAsync(conn: any, sql: string): Promise<any[]> {
  return new Promise((resolve, reject) => {
    conn.execute({
      sqlText: sql,
      complete: (err: any, _stmt: any, rows: any[]) => {
        if (err) reject(err);
        else resolve(rows);
      },
    });
  });
}

function connectAsync(conn: any): Promise<void> {
  return new Promise((resolve, reject) => {
    conn.connect((err: any) => {
      if (err) reject(err);
      else resolve();
    });
  });
}

export async function loadSnowflake(
  table: DataTable,
  dest: DestinationConfig,
  _baseDir: string,
): Promise<void> {
  const conn = snowflake.createConnection({
    account: dest.config.account as string,
    username: dest.config.user as string,
    password: dest.config.password as string,
    warehouse: dest.config.warehouse as string,
    database: dest.config.database as string,
    schema: (dest.config.schema as string) ?? "PUBLIC",
    role: dest.config.role as string,
  });

  await connectAsync(conn);
  try {
    const targetTable = ((dest.config.table as string) ?? dest.name).toUpperCase();

    if (dest.mode === "full_refresh") {
      await execAsync(conn, `DROP TABLE IF EXISTS ${targetTable}`);
    }

    // Create table
    const colDefs = table.columns
      .map((c) => `"${c.name.toUpperCase()}" VARCHAR`)
      .join(", ");
    await execAsync(
      conn,
      `CREATE TABLE IF NOT EXISTS ${targetTable} (${colDefs})`,
    );

    // Insert in batches
    const colNames = table.columns.map((c) => c.name);
    for (let i = 0; i < table.rows.length; i += dest.batch_size) {
      const batch = table.rows.slice(i, i + dest.batch_size);
      for (const row of batch) {
        const values = colNames
          .map((col) => {
            const v = row[col];
            return v === null || v === undefined ? "NULL" : `'${String(v).replace(/'/g, "''")}'`;
          })
          .join(", ");
        await execAsync(
          conn,
          `INSERT INTO ${targetTable} VALUES (${values})`,
        );
      }
    }

    console.log(
      `Loaded ${table.rows.length} rows to Snowflake destination '${dest.name}' (${targetTable})`,
    );
  } finally {
    conn.destroy(() => {});
  }
}
